cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1527,4 +1628,26 @@ private void emitTimeMetrics(String namespace, String appName, String metricName
     long timeTaken = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
     metricsCollectionService.getContext(tags).gauge(metricName, timeTaken);
   }
+
+  public void markAppsAsLatest(List<ApplicationId> appIds) throws Exception {
+    // Validate the appIds do not contain duplicates, i.e. we are not trying to mark
+    // multiple versions of the same application as latest
+    Set<String> validated = new HashSet<>();
+    for (ApplicationId appId: appIds) {
+      String appName = appId.getApplication();
+      if (validated.contains(appName)) {
+        throw new Exception(String.format(
```

Comment:
In general, there is basically never a good reason to throw Exception, something more specific should be used. In this particular case, somebody providing multiple versions of the same app should result in a 400 bad request, as it is never valid input and will always fail.

Though do note that the current logic will incorrectly error about applications from different namespaces with the same name.

Reply:
Updated the method to take namespaceId and appsMarkLatestRequest as arguments.